File: API/src/models/invoice.ts
import { Model, DataTypes, Sequelize } from 'sequelize';

interface InvoiceAttributes {
    invoiceNo: string;
    companyLogo?: Buffer;
    companyName: string;
    companyEmail: string;
    companyPhone: string;
    companyAddress: string;
    signatureImage?: Buffer;
    signatureDate?: Date;
    issueDate: Date;
    dueDate: Date;
    notes?: string;
    termsAndConditions?: string;
    userId: number;
    clientId: number;
    templateId?: number;
    subtotal: number;
    taxRate: number;
    currency: string;
    status: string;
    tax: number;
    total: number;
}

export default (sequelize: Sequelize) => {
    class Invoice extends Model<InvoiceAttributes> implements InvoiceAttributes {
        public invoiceNo!: string;
        public companyLogo?: Buffer;
        public companyName!: string;
        public companyEmail!: string;
        public companyPhone!: string;
        public companyAddress!: string;
        public signatureImage?: Buffer;
        public signatureDate?: Date;
        public issueDate!: Date;
        public dueDate!: Date;
        public notes?: string;
        public termsAndConditions?: string;
        public userId!: number;
        public clientId!: number;
        public templateId?: number;
        public subtotal!: number;
        public taxRate!: number;
        public currency!: string;
        public status!: string;
        public tax!: number;
        public total!: number;

        static associate(models: any) {
            // Association with Client
            Invoice.belongsTo(models.Client, { 
                foreignKey: 'clientId',
                as: 'client'
            });
            
            // Association with InvoiceItem
            Invoice.hasMany(models.InvoiceItem, {
                foreignKey: 'invoiceId',
                as: 'invoiceItems'
            });
        }
    }

    Invoice.init({
        invoiceNo: DataTypes.STRING,
        companyLogo: DataTypes.BLOB('medium'),
        companyName: DataTypes.STRING,
        companyEmail: DataTypes.STRING,
        companyPhone: DataTypes.STRING,
        companyAddress: DataTypes.STRING,
        signatureImage: DataTypes.BLOB('medium'),
        signatureDate: DataTypes.DATE,
        issueDate: DataTypes.DATE,
        dueDate: DataTypes.DATE,
        notes: DataTypes.TEXT,
        termsAndConditions: DataTypes.TEXT,
        userId: DataTypes.INTEGER,
        clientId: DataTypes.INTEGER,
        templateId: DataTypes.INTEGER,
        subtotal: DataTypes.FLOAT,
        taxRate: DataTypes.FLOAT,
        currency: DataTypes.STRING,
        status: DataTypes.STRING,
        tax: DataTypes.FLOAT,
        total: DataTypes.FLOAT
    }, {
        sequelize,
        modelName: 'Invoice',
        tableName: 'invoices'
    });

    return Invoice;
};

File: DotNet_API/DataModels/Invoice.cs
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;

namespace DotNet_API.DataModels
{
    [Table("invoices")]
    public class Invoice
    {
        [Key]
        [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
        public int Id { get; set; }

        [Required]
        [MaxLength(255)]
        public string InvoiceNo { get; set; } = string.Empty;

        public byte[]? CompanyLogo { get; set; }

        [Required]
        [MaxLength(255)]
        public string CompanyName { get; set; } = string.Empty;

        [Required]
        [EmailAddress]
        [MaxLength(255)]
        public string CompanyEmail { get; set; } = string.Empty;

        [Required]
        [Phone]
        [MaxLength(50)]
        public string CompanyPhone { get; set; } = string.Empty;

        [Required]
        [MaxLength(500)]
        public string CompanyAddress { get; set; } = string.Empty;

        public byte[]? SignatureImage { get; set; }

        public DateTime? SignatureDate { get; set; }

        [Required]
        public DateTime IssueDate { get; set; }

        [Required]
        public DateTime DueDate { get; set; }

        public string? Notes { get; set; }

        public string? TermsAndConditions { get; set; }

        [Required]
        public int UserId { get; set; }

        [Required]
        public int ClientId { get; set; }

        public int? TemplateId { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal Subtotal { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal TaxRate { get; set; }

        [Required]
        [MaxLength(10)]
        public string Currency { get; set; } = string.Empty;

        [Required]
        [MaxLength(50)]
        public string Status { get; set; } = string.Empty;

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal Tax { get; set; }

        [Required]
        [Column(TypeName = "decimal(18,2)")]
        public decimal Total { get; set; }

        public bool isArchived { get; set; }


        // Navigation properties
        public Client? Client { get; set; }


        public InvoiceTemplate? InvoiceTemplate { get; set; }

        public ICollection<InvoiceItem> Items { get; set; } = new List<InvoiceItem>();
    }
}
